cpp/src/arrow/util/bit-util.cc

```diff
@@ -34,13 +44,82 @@ Status BitUtil::BytesToBits(
     const std::vector<uint8_t>& bytes, std::shared_ptr<Buffer>* out) {
   int bit_length = BitUtil::BytesForBits(bytes.size());
 
-  auto buffer = std::make_shared<PoolBuffer>();
-  RETURN_NOT_OK(buffer->Resize(bit_length));
+  std::shared_ptr<MutableBuffer> buffer;
+  RETURN_NOT_OK(AllocateBuffer(default_memory_pool(), bit_length, &buffer));
+
   memset(buffer->mutable_data(), 0, bit_length);
   BytesToBits(bytes, buffer->mutable_data());
 
   *out = buffer;
   return Status::OK();
 }
 
+int64_t CountSetBits(const uint8_t* data, int64_t bit_offset, int64_t length) {
+  constexpr int64_t pop_len = sizeof(uint64_t) * 8;
+
+  int64_t count = 0;
+
+  // The first bit offset where we can use a 64-bit wide hardware popcount
+  const int64_t fast_count_start = BitUtil::RoundUp(bit_offset, pop_len);
+
+  // The number of bits until fast_count_start
+  const int64_t initial_bits = std::min(length, fast_count_start - bit_offset);
+  for (int64_t i = bit_offset; i < bit_offset + initial_bits; ++i) {
+    if (BitUtil::GetBit(data, i)) { ++count; }
+  }
+
+  const int64_t fast_counts = (length - initial_bits) / pop_len;
+
+  // Advance until the first aligned 8-byte word after the initial bits
+  const uint64_t* u64_data =
+      reinterpret_cast<const uint64_t*>(data) + fast_count_start / pop_len;
+
+  const uint64_t* end = u64_data + fast_counts;
+
+  // popcount as much as possible with the widest possible count
+  for (auto iter = u64_data; iter < end; ++iter) {
+    count += __builtin_popcountll(*iter);
```

Comment:
As this is an SSE4.2 instructions and I expect >90% of our (Python) users to have binaries compiled without `-msse4.2` but having CPUs with SSE 4.2 we might need to think about runtime detection and having a portable `CountSetBits` and faster `CountSetBits_sse42. (Note that manylinux1 and conda packages have to stick to only SSE2 support to be portable).